### PR TITLE
Array with single element inside WDL Struct doesn't break workflows [BW-678]

### PR DESCRIPTION
**Pull request by salonishah11**
Closes https://broadworkbench.atlassian.net/browse/BW-678

### CODE REVIEW

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/base/InputExpressionReassembler.scala [hunk 1]
```diff
@@ -169,17 +175,61 @@ object InputExpressionReassembler {
    */
   private def reconstructInputExprForEachEntity(mapOfEntityToEvaluatedExprMap: Map[EntityName, Try[Map[LookupExpression, JsValue]]],
                                                 parsedTree: ParseTree,
-                                                rootEntityNames: Option[Seq[EntityName]]): Map[EntityName, Try[JsValue]] = {
+                                                rootEntityNames: Option[Seq[EntityName]],
+                                                inputOption: Option[MethodInput]): Map[EntityName, Try[JsValue]] = {
+
+    def updateInputExprValueMapForObjectInput(inputExprWithEvaluatedRef: JsValue, objectFieldsFromInput: mutable.Seq[ValueTypeObjectFieldTypes]): JsObject = {
+      val updatedObject: Map[String, JsValue] = inputExprWithEvaluatedRef.asJsObject.fields.map { case (inputName, evaluatedValue) =>
```
Comment (cahrens) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/entities/base/InputExpressionReassembler.scala:185]: I'm finding this difficult to read/understand. Why does there need to be a case statement (don't we do this for all keys/values)?

Comment (salonishah11) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/entities/base/InputExpressionReassembler.scala:185]: Yes. We are iterating for all (key, value) pairs inside the map. Since the data type of fields inside a JsObject is Map[String, JsValue], we can use a case statement to identify the key and value separately. This can also be written as below, but using case statement and having explicit variable names makes it more readable. In my opinion, using `obj._1` and `obj._2` makes it harder to understand/read what is going on. We can always extract `obj._1` and `obj._2` into variables but case statement implicitly does that.
```
inputExprWithEvaluatedRef.asJsObject.fields.map { obj =>
// in this case obj._1 will be key of type String and obj._2 will be the value of type JsValue

val inputObjectField: Option[ValueTypeObjectFieldTypes] = objectFieldsFromInput.find(x => x.getFieldName == obj._1)
...
}
```

**Comment by salonishah11** [comment]
jenkins retest

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/base/InputExpressionReassembler.scala (hunk 1 above)
Comment (cahrens) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/entities/base/InputExpressionReassembler.scala:185]: Thanks! Doing operations through case statements still doesn't feel intuitive to me yet, but hopefully it will soon!

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/base/InputExpressionReassembler.scala [hunk 2]
```diff
@@ -169,17 +175,61 @@ object InputExpressionReassembler {
...
+        val inputObjectField: Option[ValueTypeObjectFieldTypes] = objectFieldsFromInput.find(x => x.getFieldName == inputName)
+
+        val updatedValue: JsValue = inputObjectField match {
+          case Some(objectField) => objectField.getFieldType.getTypeName match {
+            case TypeNameEnum.OBJECT =>
+              // if there are nested objects, recursively call this method to reach leaf fields
```
Comment (cahrens) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/entities/base/InputExpressionReassembler.scala:191]: Do you have a test case that covers this? I'm having trouble understanding the test cases.

Comment (salonishah11) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/entities/base/InputExpressionReassembler.scala:191]: Yes. [This](https://github.com/broadinstitute/rawls/blob/e7065edb153ba75f5fc47bdc436d12c67f11f15f/core/src/test/scala/org/broadinstitute/dsde/rawls/entities/local/LocalEntityProviderSpec.scala#L121-L131) test case should cover it

**Review by cahrens — APPROVED** [review]

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/base/InputExpressionReassembler.scala [hunk 3]
```diff
@@ -169,17 +175,61 @@ object InputExpressionReassembler {
    */
   private def reconstructInputExprForEachEntity(mapOfEntityToEvaluatedExprMap: Map[EntityName, Try[Map[LookupExpression, JsValue]]],
                                                 parsedTree: ParseTree,
-                                                rootEntityNames: Option[Seq[EntityName]]): Map[EntityName, Try[JsValue]] = {
+                                                rootEntityNames: Option[Seq[EntityName]],
+                                                inputOption: Option[MethodInput]): Map[EntityName, Try[JsValue]] = {
+
+    def updateInputExprValueMapForObjectInput(inputExprWithEvaluatedRef: JsValue, objectFieldsFromInput: mutable.Seq[ValueTypeObjectFieldTypes]): JsObject = {
```
Comment (cjllanwarne) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/entities/base/InputExpressionReassembler.scala:184]: Docs please!

Comment (cjllanwarne) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/entities/base/InputExpressionReassembler.scala:184]: eg point out that this is specifically for inputs which are formatted like json objects

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/base/InputExpressionReassembler.scala (hunk 2 above)
Comment (cjllanwarne) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/entities/base/InputExpressionReassembler.scala:191]: It looks like we have one test, but this kind of code is always potentially tricky. We should probably validate all possible input permutations (and possibly highlight known bugs with "we should fix this known problem" test cases.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/base/InputExpressionReassembler.scala [hunk 4]
```diff
@@ -169,17 +175,61 @@ object InputExpressionReassembler {
...
       }
 
-      entityName -> inputExprWithEvaluatedRef
+      /* In SlickExpressionEvaluator, for attribute references that evaluate to AttributeValueList type, the elements inside the list are extracted and returned.
+         As a result, for attributes that contain only 1 element, we can no longer identify whether the original value was a list or a single AttributeValue.
+         Because of this type erasure, when we are reconstructing object type inputs, List/Array with single element get incorrectly reconstructed as single AttributeValue,
+         which leads to Cromwell failing while processing inputs (https://broadworkbench.atlassian.net/browse/BW-678).
+         To fix this, once the object type inputs have been reconstructed, we iterate over each field for object input and verify for Array type inputs that the
+         evaluated value is also an Array. If it isn't we convert it into Array and updated the evaluated object.
+         Note: This approach does not fix the problem if there is single element array inside an array. This would be fixed in https://broadworkbench.atlassian.net/browse/BW-846.
```
Comment (cjllanwarne) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/entities/base/InputExpressionReassembler.scala:223]: Or you could have an equivalent, recursive `updateInputExprValueMapForArrayInput`, presumably?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/base/InputExpressionReassembler.scala (hunk 3 above)
Comment (cjllanwarne) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/entities/base/InputExpressionReassembler.scala:184]: And that it is called further down this method for generating the `updatedEvaluatedRefMap` if the input is an object, in order to array-ify any not-yet-detected array values

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/base/InputExpressionReassembler.scala (hunk 2 above)
Comment (cjllanwarne) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/entities/base/InputExpressionReassembler.scala:191]: By permutations I mean things like:

Array types:
- Empty arrays
- Single element arrays
- Multi-element arrays

In places like:
- As top-level inputs
- Fields in objects
- Fields in nested objects
- Values in arrays
- Values in nested arrays

etc

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/base/InputExpressionReassembler.scala (hunk 4 above)
Comment (salonishah11) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/entities/base/InputExpressionReassembler.scala:223]: Hmm, interesting point. Just so that I am understanding correctly, if the input type is an Array, we recursively call `updateInputExprValueMapForArrayInput` until the type inside an Array is not an Array. I think it would be possible since the fieldType of nested Array is an Array 🤔

Comment (cjllanwarne) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/entities/base/InputExpressionReassembler.scala:223]: I'm not sure, I was thinking you could just have an equivalent function to the one that looks inside objects for looking inside arrays... but I'm not sure. I wasn't suggesting _doing_ this, more that it's another option when we do come to look at this story again

Comment (salonishah11) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/entities/base/InputExpressionReassembler.scala:223]: Ah I see what you mean. Yeah looking inside Array inputs might be tricky especially for deeper nested arrays. I would hope if we look at this bug again, we would solve the problem for real by fixing the type erasure.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/base/InputExpressionReassembler.scala (hunk 2 above)
Comment (salonishah11) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/entities/base/InputExpressionReassembler.scala:191]: Sounds good. I will add more permutations. I think multi-element array tests are already present. I will add the rest.

Comment (salonishah11) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/entities/base/InputExpressionReassembler.scala:191]: I am still working on the tests, but I was wondering how do we test the test cases that we know fail? For example, Single element array as values in array or values in nested arrays will fail. How do we assert that?

Comment (salonishah11) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/entities/base/InputExpressionReassembler.scala:191]: I have added test cases from the permutation list that was not previously present. Test cases for different array types as top-level inputs were already present.

**Review by cjllanwarne — APPROVED** [review]

File: core/src/test/scala/org/broadinstitute/dsde/rawls/entities/local/LocalEntityProviderSpec.scala [hunk 5]
```diff
@@ -82,6 +82,18 @@ class LocalEntityProviderSpec extends AnyWordSpecLike with Matchers with TestDri
         Map(sampleSet2.name -> Seq(SubmissionValidationValue(Some(AttributeValueEmptyList), None, intArrayNameWithWfName)))
     }
 
+    "resolve empty lists into empty Array in nested WDL Struct" in withConfigData {
+      val context = workspace
```
Comment (cahrens) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/entities/local/LocalEntityProviderSpec.scala:86]: Nit: you could refactor these test points to share the common code. The inputs and expected outputs vary, but everything else is the same.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/MethodConfigTestSupport.scala [hunk 6]
```diff
@@ -325,7 +439,8 @@ trait MethodConfigTestSupport {
       AttributeName.withDefaultNS("samples") -> AttributeEntityReferenceList(Seq(
         sampleGood.toReference,
         sampleGood2.toReference
-      ))
+      )),
+      AttributeName.withDefaultNS("rawJsonDoubleArray") -> AttributeValueRawJson( "[[0,1,2],[3,4,5]]".parseJson )
```
Comment (cahrens) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/MethodConfigTestSupport.scala:443]: Why did you add this?

Comment (salonishah11) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/MethodConfigTestSupport.scala:443]: This was added for test case where we have multi-element array inside an array

File: core/src/test/scala/org/broadinstitute/dsde/rawls/entities/local/LocalEntityProviderSpec.scala (hunk 5 above)
Comment (salonishah11) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/entities/local/LocalEntityProviderSpec.scala:86]: Since this is test code, I am inclined to leave it like this as in my opinion time involved in refactoring outweighs the benefits

Comment (salonishah11) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/entities/local/LocalEntityProviderSpec.scala:86]: Having said that, if you do feel strongly, I can refactor them :)

Comment (cahrens) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/entities/local/LocalEntityProviderSpec.scala:86]: No, I don't feel strongly about it-- that's why I prefaced it with "Nit".

In Python unit tests, there was a nice way to annotate a test method as something that should run with a list of inputs/expected outputs, and the test harness then ran the test point with each set of inputs/outputs individually. It was a very convenient way to keep the tests concise and maintainable. Are you aware of anything like this in Scala?

Comment (salonishah11) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/entities/local/LocalEntityProviderSpec.scala:86]: I am not sure. I will look it up 👍

Comment (salonishah11) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/entities/local/LocalEntityProviderSpec.scala:86]: I was able to find [this](https://github.com/broadinstitute/rawls/blob/7bc2274032277fd9605495fbeb4c7a97daefc370/core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/SubmissionApiServiceSpec.scala#L837-L872) pattern. Is this what you had in mind?

Comment (cahrens) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/entities/local/LocalEntityProviderSpec.scala:86]: Yes, but I was wondering if there were language constructs/libraries to make this easier. For example, the ddt library in Python as shown below.

Don't go searching if you don't know… I was just curious if you were aware of anything like this.

````python
    @data(
        ['Alewife', 'Alewife', [['Red Line']]],
        ['Shawmut', 'Quincy Adams', [['Red Line']]],
        ['Downtown Crossing', 'Alewife', [['Red Line']]],
        ['Downtown Crossing', 'Haymarket', [['Orange Line']]],
        ['Ashmont', 'Milton', [['Mattapan Trolley']]],
        ['Haymarket', 'Prudential', [['Green Line E']]],
        ['Park Street', 'North Station', [['Green Line C'], ['Green Line E']]],
        ['Ruggles', 'Wonderland', [['Orange Line', 'Blue Line']]],
        ['South Station', 'State', [['Red Line', 'Orange Line']]],
        ['Copley', 'Chinatown', [['Green Line C', 'Orange Line'],
                                 ['Green Line E', 'Orange Line']]],
        ['Fenway', 'North Station', [['Green Line D', 'Green Line C'],
                                     ['Green Line D', 'Green Line E']]],
        ['Mattapan', 'Roxbury Crossing', [['Mattapan Trolley', 'Red Line', 'Orange Line']]],
        ['Broadway', 'Aquarium', [['Red Line', 'Orange Line', 'Blue Line'],
                                  ['Red Line', 'Green Line C', 'Blue Line'],
                                  ['Red Line', 'Green Line D', 'Blue Line'],
                                  ['Red Line', 'Green Line E', 'Blue Line']]],
        ['Fenway', 'Assembly', [['Green Line D', 'Green Line C', 'Orange Line'],
                                ['Green Line D', 'Green Line E', 'Orange Line'],
                                ['Green Line D', 'Blue Line', 'Orange Line']]]
    )
    @unpack
    def test_get_route(self, source, destination, possible_routes):
        """
        Tests the get_route method (Question 3). If multiple routes are possible,
        makes sure that the route returned is in the list of possible routes.
        """
        route = self.subway.get_route(source, destination)
        self.assertIn(
            route, possible_routes,
            'Route incorrect for stops {source} -> {destination}. Expected one of {possible_routes}, but got {route}.'.format(
                source=source, destination=destination, possible_routes=possible_routes, route=route
            )
        )

Comment (salonishah11) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/entities/local/LocalEntityProviderSpec.scala:86]: No I am not aware of such a library. I can look into it outside of this PR when I have some time (out of curiosity).

Comment (cahrens) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/entities/local/LocalEntityProviderSpec.scala:86]: I've been Googling a bit and haven't found anything yet. It's clearly just "syntactic sugar", but I found that it reduced my tendency to copy/paste test code, and it makes the tests easier to understand because the team becomes familiar with a single syntax, vs. everyone rolling their own syntax for this sort of thing.
